### PR TITLE
ccl/backupccl,ccl/importccl: disable merges during IMPORT and RESTORE

### LINKED ISSUE #29268: storage: use gossip to avoid merging away restore/import's splits
Restore/import splits the keyspace and doesn't want the merge queue to merge these splits away. In the absence of "sticky bits" (and perhaps even after we introduce sticky bits) we want to let the merge queue know to respect these (temporary) splits.

The suggestion is for the restore/import job to regularly gossip a key whose value contains a key prefix on which merges are not to be carried out, with a relatively short TTL (minutes).

The merge queue looks out for these keys and will not attempt to merge two ranges when either is marked as protected.

### PR DESCRIPTION
**Pull request by petermattis**
Add a facility for periodically gossiping keys which disable range
merges for a set of tables. This is used by IMPORT and RESTORE to
disable range merges on tables being imported / restored. This is done
to prevent the merge queue from fighting against the splitting performed
by IMPORT and RESTORE.

Fixes #29268

Release note: None

### CODE REVIEW

**Review by benesch — APPROVED** [review]
:lgtm_strong: :tada: thank you!

> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-:-LRr2Lkt5g4ngrR8FsFR:btlkf9r)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/merge_queue.go, line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRr0bwe2N7yOu1bCv48:-LRr0bwe2N7yOu1bCv49:bv7ea0) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eba6d82d73eb31c1a7103731e141c3e836ba43/pkg/storage/merge_queue.go#L127)):*
> ```Go
> }
> 
> func (mq *mergeQueue) mergesDisabled(desc *roachpb.RangeDescriptor) bool {
> ```

slight preference for `mergesDisabledForRange` or similar to disambiguate from the cluster-wide case
___
*[pkg/storage/merge_queue.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRr0s3q-T1cP6VmotOB:-LRr0s3r5GS6GPvm-FHr:bu5tkzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eba6d82d73eb31c1a7103731e141c3e836ba43/pkg/storage/merge_queue.go#L167)):*
> ```Go
> 		return false, 0
> 	}
> ```

nit: this splits the computation of sizeRatio from its usage. I think this should go after NeedsSplit and before sizeRatio.

**Review by dt — APPROVED** [review]
:lgtm:

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrms7n-JyN_l-SxZV3:-LRrms7o6s9twzUun1Lu:bf9v7m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L245)):*
> ```Go
> 						"cr.store.queue.merge.process.success": 20,
> 						"cr.store.queue.merge.process.failure": 20,
> 					})
> ```

I'm not sure if these thresholds are right as I'm still seeing a handful of merges during restore. PTAL at how this mechanism works.
___
*[pkg/cmd/roachtest/restore.go, line 272 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrn0Vr4FWvjh0e-sMZ:-LRrn0Vs1Z6qSV02Vati:b-hctqqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L272)):*
> ```Go
> // metric does not exceed some threshold during a test. For example, the
> // restore and import tests verify that the range merge queue is inactive.
> func verifyMetrics(ctx context.Context, c *cluster, m map[string]float64) error {
> ```

@mrtracy Inspired by `verifyTxnPerSecond`. PTAL and tell me if I'm doing anything wrong.

**Review by mrtracy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrwazhA4su_wjGbyZF:-LRrwazhA4su_wjGbyZG:b-ypxj0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L302)):*
> ```Go
> 
> 		now := timeutil.Now()
> 		request.StartNanos = now.Add(-sample).UnixNano()
> ```

What's the specific value in querying this as-soon-as-possible?  I would suggest querying for a different range, something like [now-sample*3,  now-sample], to ensure that you always get a few samples in the near past (and also don't run into the 'cannot query time series in the future' issue you're seeing here).  The time series data is always historical, so you will still detect the same failures you were detecting before, you'll just detect them ten seconds later in the test.
___
*[pkg/cmd/roachtest/restore.go, line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrxahX4yZQ5XXTK6fE:-LRrxahX4yZQ5XXTK6fF:bkxzgmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L324)):*
> ```Go
> 			value := data[n-1].Value
> 			if value >= limit {
> 				return fmt.Errorf("%s: %.1f >= %.1f", name, value, limit)
> ```

Might want to output the timestamp of the failed sample to correlate it with logs.

**Review by mrtracy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrwazhA4su_wjGbyZF:-LRrxhkM1fB8ydE_zqrZ:b-9qy7e0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L302)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

What's the specific value in querying this as-soon-as-possible?  I would suggest querying for a different range, something like [now-sample*3,  now-sample], to ensure that you always get a few samples in the near past (and also don't run into the 'cannot query time series in the future' issue you're seeing here).  The time series data is always historical, so you will still detect the same failures you were detecting before, you'll just detect them ten seconds later in the test.
</blockquote></details>

s/as-soon-as-possible/as-*recently*-as-possible/

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrms7n-JyN_l-SxZV3:-LRrsz6C54n5tif4Tokb:b43wpk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L245)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not sure if these thresholds are right as I'm still seeing a handful of merges during restore. PTAL at how this mechanism works.
</blockquote></details>

Ok, I figured out the reason for the handful of merges. My mechanism to disable range merges was only checking the start key of a range. The merge queue was periodically picking up the range immediately before the table being restored and seeing it as a candidate for merging and queueing the range. When the queue went to process the range it would decide not to do anything. But despite not doing anything the merge queue success metric would be bumped because that is how queue metrics work.

Something curious is that I wasn't seeing the log message from the merge queue I was expecting. Specifically, I expected the ranges not to be merged because the merged range would require a split. I'm going to dig into this further.
___
*[pkg/storage/merge_queue.go, line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRruAJHEsDocSmZQTFy:-LRruAJHEsDocSmZQTFz:b-id1cx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/storage/merge_queue.go#L143)):*
> ```Go
> 	}
> 	_, err = mq.gossip.GetInfo(gossip.MakeTableDisableMergesKey(uint32(tableID2)))
> 	return err == nil
> ```

@benesch The logic changed here. Now I'm checking both the start key and the end key so that we don't queue a range if merges are disabled on it or on its right sibling.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrms7n-JyN_l-SxZV3:-LRrzqUdEvGDkV5p6NOv:b-2brb0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L245)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok, I figured out the reason for the handful of merges. My mechanism to disable range merges was only checking the start key of a range. The merge queue was periodically picking up the range immediately before the table being restored and seeing it as a candidate for merging and queueing the range. When the queue went to process the range it would decide not to do anything. But despite not doing anything the merge queue success metric would be bumped because that is how queue metrics work.

Something curious is that I wasn't seeing the log message from the merge queue I was expecting. Specifically, I expected the ranges not to be merged because the merged range would require a split. I'm going to dig into this further.
</blockquote></details>

@benesch I added a bit of logic to the merge queue to understand the weirdness I'm seeing and disabled a portion of this PR. The log message says:

```
I181121 20:31:38.268606 1751 storage/merge_queue.go:272  [n1,merge,s1,r20/1:/Table/{23-53}] merging ranges: /Table/{23-53} + /{Table/53-Max} -> /{Table/23-Max}
```

This is output immediately after the call to `shouldSplitRange`. But this is super bizarre because `/Table/53` is the table being restored. 

Oh, I see what the problem is. We've allocated ID 53 for the table being restored, but the descriptor isn't written until the restore is finished. So `sysCfg.NeedSplit` doesn't see a need for a split at `/Table/53`. I'm going to file an issue about this. It is somewhat surprising that a table being restored doesn't show up in the admin UI, and it also has an effect here.
___
*[pkg/cmd/roachtest/restore.go, line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrwazhA4su_wjGbyZF:-LRryWuV6K62e48dfHLu:b2qtwnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L302)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

s/as-soon-as-possible/as-*recently*-as-possible/
</blockquote></details>

How will `now-sample*3` avoid the `cannot query time series in the future` issue? Is that the only change you're suggesting here. There is no particular need to have the most up to date data, I just want the last time series value for these metrics. Tell me what code to type in order to achieve that, sensei. 
___
*[pkg/cmd/roachtest/restore.go, line 324 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrxahX4yZQ5XXTK6fE:-LRrzM2u1i-tuhnk0QZw:b52iz3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L324)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Might want to output the timestamp of the failed sample to correlate it with logs.
</blockquote></details>

Done (or it will be when I push, I'm pulling on a thread right now that wants to unravel).

**Review by mrtracy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrwazhA4su_wjGbyZF:-LRs3PfiBGYioFezec_g:b-wj7a6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L302)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How will `now-sample*3` avoid the `cannot query time series in the future` issue? Is that the only change you're suggesting here. There is no particular need to have the most up to date data, I just want the last time series value for these metrics. Tell me what code to type in order to achieve that, sensei. 
</blockquote></details>

Well, the most efficient option in terms of queries would be to wait until the end of the test, then query the time series for the entire duration of the test and verify every individual timestamp.  However, it appears your intention here is to stop the test as soon as this condition fails.

The "cannot query in the future" is actually occurring because you cannot query for the current time period while it is still open (to avoid incomplete datapoints, where not all servers have reported). So, it's not really safe to query *only* ten seconds in the past, you need to have a start time earlier than that.

My suggestion is just to grab the last few datapoints by querying starting from three sample periods in the past, instead of just one; this will definitely return datapoints and will not run into this particular error. From a query perspective this is not more expensive than querying a single datapoint.

So in clear terms, if you use this line:

`request.StartNanos = now.Add(-sample*3).UnixNano()`

I think this will work as-is and the code should never encounter the "time series in the past" error; in fact, you should remove that check because it would indicate an actual problem.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrms7n-JyN_l-SxZV3:-LRs519uBpPoLEbwcf82:b-a01pvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L245)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@benesch I added a bit of logic to the merge queue to understand the weirdness I'm seeing and disabled a portion of this PR. The log message says:

```
I181121 20:31:38.268606 1751 storage/merge_queue.go:272  [n1,merge,s1,r20/1:/Table/{23-53}] merging ranges: /Table/{23-53} + /{Table/53-Max} -> /{Table/23-Max}
```

This is output immediately after the call to `shouldSplitRange`. But this is super bizarre because `/Table/53` is the table being restored. 

Oh, I see what the problem is. We've allocated ID 53 for the table being restored, but the descriptor isn't written until the restore is finished. So `sysCfg.NeedSplit` doesn't see a need for a split at `/Table/53`. I'm going to file an issue about this. It is somewhat surprising that a table being restored doesn't show up in the admin UI, and it also has an effect here.
</blockquote></details>

Holding off on writing the descriptor until the end was an intentional decision -- we were pretty uncomfortable with AddSSTable's relatively unsafe interaction with anyone else trying to operate in the same keys space, so not writing a descriptor was how we keep SQL/the adminui/etc, from having any reason to be in our key range.

File: pkg/storage/merge_queue.go
Comment (tbg) [inline pkg/storage/merge_queue.go]: Couldn't the range span multiple tables? I think this means you have to look at the range tableID1..tableID2

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/merge_queue.go, line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRqxaXkod27PC2VwaVL-r4-128:-LRs9nBtAWzm_trSALet:b-8sog3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a931344f5a8dfb028d3681a14810012e79fc579c/pkg/storage/merge_queue.go#L128)):*
<details><summary><i>Previously, tbg (Tobias Schottdorf) wrote…</i></summary><blockquote>

Couldn't the range span multiple tables? I think this means you have to look at the range tableID1..tableID2
</blockquote></details>

Yes, the range can span multiple tables, but the `sysCfg.NeedsSplit` will capture that. The only case here that is interesting is when restore or import have created a table that isn't yet in `system.descriptors`. I believe that makes it fine to only check the table IDs in the start and end key, and not everything in between.

**Review by benesch — APPROVED** [review]
> Reviewed 1 of 1 files at r2, 1 of 2 files at r3, 4 of 4 files at r4, 1 of 2 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrms7n-JyN_l-SxZV3:-LRtJAEaDQI7F2w1E6wk:bhy9cyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L245)):*
> Ok, I figured out the reason for the handful of merges. My mechanism to disable range merges was only checking the start key of a range. The merge queue was periodically picking up the range immediately before the table being restored and seeing it as a candidate for merging and queueing the range. When the queue went to process the range it would decide not to do anything. But despite not doing anything the merge queue success metric would be bumped because that is how queue metrics work.

We need to fix this more generally at the metrics level. There are various other queues that can nope out of processing a replica but still have that counted as a successful action. The replica GC queue comes to mind.

Your solution here LGTM though.
___
*[pkg/cmd/roachtest/restore.go, line 328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRtIDEl-eaRElfO5xDc:-LRtIDEm9lwmJGiZZFdG:b4m9zo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aef3381a89445374f73999f302d74ac60e9bd48/pkg/cmd/roachtest/restore.go#L328)):*
> ```Go
> 		}
> 	}
> }
> ```

I didn't review this, just FYI. Figured @mrtracy had it covered.
___
*[pkg/storage/merge_queue.go, line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRqxaXkod27PC2VwaVL-r4-128:-LRtJsVhBRz8r-ovUqlu:b-pbdqz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a931344f5a8dfb028d3681a14810012e79fc579c/pkg/storage/merge_queue.go#L128)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, the range can span multiple tables, but the `sysCfg.NeedsSplit` will capture that. The only case here that is interesting is when restore or import have created a table that isn't yet in `system.descriptors`. I believe that makes it fine to only check the table IDs in the start and end key, and not everything in between.
</blockquote></details>

Agreed.
___
*[pkg/storage/merge_queue.go, line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRruAJHEsDocSmZQTFy:-LRtJoRnB2Yp_Chy6FBW:bb6ywmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/storage/merge_queue.go#L143)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@benesch The logic changed here. Now I'm checking both the start key and the end key so that we don't queue a range if merges are disabled on it or on its right sibling.
</blockquote></details>

LGTM
___
*[pkg/storage/merge_queue_test.go, line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRtJbuv5sBbt0n35XrM:-LRtJbuv5sBbt0n35XrN:bl68oa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a931344f5a8dfb028d3681a14810012e79fc579c/pkg/storage/merge_queue_test.go#L168)):*
> ```Go
> 		{
> 			startKey: tableKey(4),
> 			endKey:   append(tableKey(3), 'z'),
> ```

Isn't this interval inverted? Did you want `append(tableKey(4), 'z'`)?

**Comment by spencerkimball** [comment]
I was under the impression that when splitting by load or Admin intention, we noted that fact in the range descriptor to prevent merging for some period of time. Do we not prevent the merge queue from attempting to merge load based splits?

**Comment by petermattis** [comment]
> I was under the impression that when splitting by load or Admin intention, we noted that fact in the range descriptor to prevent merging for some period of time. Do we not prevent the merge queue from attempting to merge load based splits?

Yes, the merge queue has logic to avoid merging ranges if load-based splitting would split them right away. It does this by assuming the load on the RHS range is the same as the load on the LHS range (it only knows the load on the LHS range because the RHS might be remote).

We don't currently have any mechanism to prevent the merge queue from merging ranges split manually.

**Comment by spencerkimball** [comment]
The lack of a mechanism to introduce hysteresis here feels short sighted. As soon as load slacks on a busy LBS range, it’ll get merged. Seems we should wait a day for that as load is somewhat (at least) likely to reappear on formerly busy ranges.

It feels overly complex to introduce a gossip based merge-freezing mechanism when something more fundamental at the range descriptor level feels necessary for load based splitting, and would handle both cases with more deliberate intention.

**Comment by petermattis** [comment]
@benesch Care to weigh in here? I know you considered placing some marker in the RangeDescriptor, but I don't recall why that wasn't implemented.

**Comment by tbg** [comment]
Adding this to the range descriptor is the alternative to this (and probably the solution for 2.2), but it could be a little more involved than anticipated. Adding fields to protos which we use via CPut is difficult because old versions will (perhaps?) drop the new field on the floor and get bogus mismatches (failed replica changes). I think we've been burned by this in the past more than once.

On the flip side, this change leaves no permanent trace and can be kicked out at any time, so it neither locks us into any particular decision nor is it likely to cause trouble (including in backports).

**Comment by spencerkimball** [comment]
@tbg, I don’t fully understand he point about CPut. Wouldn’t the value of whatever sub proto we might add – to indicate reason for a split – just be transmitted with the expected value on the next CPut?

**Comment by tbg** [comment]
@spencerkimball imagine you're a node which doesn't know about this new field, and you read a range descriptor which has the field. You get `RangeDescriptor{Stuff}` instead of `RangeDescriptor{Stuff, NewField: NewStuff}` (ignoring `XXX_Unrecognized` which may save the day but I'm not sure about that). When you go and do a `CPut`, you'll serialize `RangeDescriptor{Stuff}` instead of `RangeDescriptor{Stuff, NewField: NewStuff}` and the `CPut` fails.

This can all be made work (and maybe it'll "just work"), but it's more work than this change and I don't want to introduce any more bugs right now. I agree that there's a shortcoming with respect to load based splitting, but the current priority is getting our fundamental stability problems under control (#31409), which is what ultimately motivated this change. 

Perhaps the concern is overblown and going the other route right away would've made sense, which is something you could explore if you have the bandwidth but note that you'd need mixed-version testing and must not introduce a cluster setting (if this needs to be backported, which maybe isn't a requirement with this PR in).

**Comment by benesch** [comment]
I added a field to the RangeDescriptor proto without too much trouble here: https://github.com/cockroachdb/cockroach/pull/28071/commits/ce936508d150f1caa5ca94fb3a66ba84e4555e3e. The trick is to avoid the set-but-zero state. New fields need to be either entirely unset or set to non-zero value.

I agree that we should commit this fix in the short time. I also agree that right solution involves storing persistent data somewhere—maybe the range descriptor, maybe a range local key—but I think we want to send any change like that through a lightweight RFC process.

**Review by petermattis — COMMENTED** [review]
Ok, this should be good to go.

@spencerkimball Did the response from @tbg and @benesch mitigate your concerns?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-:-LSGBSsB7K-v4WlxMR39:bf6l4mf)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 302 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRrwazhA4su_wjGbyZF:-LSGAehL27HXDzmPrzKJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f83dc584b37d9c25099f4f44c01ffe04f447037/pkg/cmd/roachtest/restore.go#L302)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Well, the most efficient option in terms of queries would be to wait until the end of the test, then query the time series for the entire duration of the test and verify every individual timestamp.  However, it appears your intention here is to stop the test as soon as this condition fails.

The "cannot query in the future" is actually occurring because you cannot query for the current time period while it is still open (to avoid incomplete datapoints, where not all servers have reported). So, it's not really safe to query *only* ten seconds in the past, you need to have a start time earlier than that.

My suggestion is just to grab the last few datapoints by querying starting from three sample periods in the past, instead of just one; this will definitely return datapoints and will not run into this particular error. From a query perspective this is not more expensive than querying a single datapoint.

So in clear terms, if you use this line:

`request.StartNanos = now.Add(-sample*3).UnixNano()`

I think this will work as-is and the code should never encounter the "time series in the past" error; in fact, you should remove that check because it would indicate an actual problem.
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/restore.go, line 242 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LSGAHFZDq2Dl_sYXwyn:-LSGAHFZDq2Dl_sYXwyo:bxa6cbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f7b4e5fb56001c32dd9bc6297e22f60855fbaf4/pkg/cmd/roachtest/restore.go#L242)):*
> ```Go
> 
> 				// TODO(peter): This currently causes the test to fail because we see a
> 				// flurry of valid merges when the restore finishes.
> ```

I had to disable my nifty merge check in this test because it would fire incorrectly at the end of the test. Once the restore is finished merges are unblocked and the merge queue would merrily merge a number of ranges. I need to think what can be done about this, but I don't want the fix to hold up this PR.
___
*[pkg/storage/merge_queue.go, line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRqxaXkod27PC2VwaVL-r4-128:-LSGB8_N4yIq41rf6tXC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a931344f5a8dfb028d3681a14810012e79fc579c/pkg/storage/merge_queue.go#L128)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Agreed.
</blockquote></details>

Done.
___
*[pkg/storage/merge_queue_test.go, line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LRtJbuv5sBbt0n35XrM:-LSGAo1f5SjPNOi1ZWXr:b-zh2yfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a931344f5a8dfb028d3681a14810012e79fc579c/pkg/storage/merge_queue_test.go#L168)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Isn't this interval inverted? Did you want `append(tableKey(4), 'z'`)?
</blockquote></details>

Yep. Fixed.

**Comment by spencerkimball** [comment]
Yes. Just wanted to be sure that the long term fix is on the horizon.

**Review by benesch — APPROVED** [review]
> Reviewed 3 of 3 files at r9, 2 of 2 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32538)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/restore.go, line 242 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32538#-LSGAHFZDq2Dl_sYXwyn:-LSGR3GWCIrhya_uZlAN:b-yuk9om) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f7b4e5fb56001c32dd9bc6297e22f60855fbaf4/pkg/cmd/roachtest/restore.go#L242)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I had to disable my nifty merge check in this test because it would fire incorrectly at the end of the test. Once the restore is finished merges are unblocked and the merge queue would merrily merge a number of ranges. I need to think what can be done about this, but I don't want the fix to hold up this PR.
</blockquote></details>

Why is the restore creating ranges that should be merged away? o_O Perhaps the store dump has some bad splits that restore is faithfully recreating, in which case we could regen the store dump on a cluster with merges enabled.

**Comment by petermattis** [comment]
> Why is the restore creating ranges that should be merged away? o_O Perhaps the store dump has some bad splits that restore is faithfully recreating, in which case we could regen the store dump on a cluster with merges enabled.

Unfortunately, I nuked that cluster. I'm sure this would reproduce if you ran `restore2TB` yourself. 😄

**Comment by petermattis** [comment]
bors r=benesch,dt

**Comment by petermattis** [comment]
Known CI failure: https://github.com/cockroachdb/cockroach/issues/31974

bors r=benesch,dt

**Comment by benesch** [comment]
The Bors batch timed out. Sigh. Bors, why don't you have any retry loops?

bors r+

**Pull request merged**
